Mathlib/Algebra/Group/AddChar.lean

```diff
@@ -31,170 +30,232 @@ see `Mathlib.NumberTheory.LegendreSymbol.AddCharacter`.
 
-open Multiplicative
-
-/-- Define coercion to a function so that it includes the move from `A` to `Multiplicative A`.
-After we have proved the API lemmas below, we don't need to worry about writing `ofAdd a`
-when we want to apply an additive character. -/
-instance instFunLike : FunLike (AddChar A M) A M :=
-  inferInstanceAs (FunLike (Multiplicative A →* M) A M)
-#noalign add_char.has_coe_to_fun
+-- this instance was a bad idea and conflicted with `instFunLike` above
+#noalign add_char.monoid_hom_class
 
-theorem coe_to_fun_apply (ψ : AddChar A M) (a : A) : ψ a = ψ.toMonoidHom (ofAdd a) :=
+@[simp] lemma toMonoidHom_apply (ψ : AddChar A M) (a : Multiplicative A) : ψ.toMonoidHom a = ψ a :=
   rfl
-#align add_char.coe_to_fun_apply AddChar.coe_to_fun_apply
+#align add_char.coe_to_fun_apply AddChar.toMonoidHom_apply
 
--- Porting note: added
-theorem mul_apply (ψ φ : AddChar A M) (a : A) : (ψ * φ) a = ψ a * φ a :=
-  rfl
+/-- An additive character maps multiples by natural numbers to powers. -/
+@[simp] lemma map_nsmul_pow (ψ : AddChar A M) (n : ℕ) (x : A) : ψ (n • x) = ψ x ^ n :=
+  ψ.toMonoidHom.map_pow x n
+#align add_char.map_nsmul_pow AddChar.map_nsmul_pow
+
+variable (A M) in
+/-- Additive characters `A → M` are the same thing as monoid homomorphisms from `Multiplicative A`
+to `M`. -/
+def toMonoidHomEquiv : AddChar A M ≃ (Multiplicative A →* M) where
+  toFun φ := φ.toMonoidHom
+  invFun f :=
+  { toFun := f.toFun
+    map_zero_one' := f.map_one'
+    map_add_mul' := f.map_mul' }
+  left_inv _ := rfl
+  right_inv _ := rfl
+
+/-- Interpret an additive character as a monoid homomorphism. -/
+def toAddMonoidHom (φ : AddChar A M) : A →+ Additive M where
+  toFun := φ.toFun
+  map_zero' := φ.map_zero_one'
+  map_add' := φ.map_add_mul'
+
+variable (A M) in
+/-- Additive characters `A → M` are the same thing as additive homomorphisms from `A` to
+`Additive M`. -/
+def toAddMonoidHomEquiv : AddChar A M ≃ (A →+ Additive M) where
+  toFun φ := φ.toAddMonoidHom
+  invFun f :=
+  { toFun := f.toFun
+    map_zero_one' := f.map_zero'
+    map_add_mul' := f.map_add' }
+  left_inv _ := rfl
+  right_inv _ := rfl
+
+/-- The trivial additive character (sending everything to `1`) is `(1 : AddChar A M).` -/
+instance instOne : One (AddChar A M) := (toMonoidHomEquiv A M).one
 
 -- Porting note: added
-@[simp]
-theorem one_apply (a : A) : (1 : AddChar A M) a = 1 := rfl
+@[simp] lemma one_apply (a : A) : (1 : AddChar A M) a = 1 := rfl
 
--- this instance was a bad idea and conflicted with `instFunLike` above
-#noalign add_char.monoid_hom_class
+instance instInhabited : Inhabited (AddChar A M) := ⟨1⟩
 
--- Porting note(https://github.com/leanprover-community/mathlib4/issues/5229): added.
-@[ext]
-theorem ext (f g : AddChar A M) (h : ∀ x : A, f x = g x) : f = g :=
-  MonoidHom.ext h
+/-- Composing a `MonoidHom` with an `AddChar` yields another `AddChar`. -/
+def monoidHom_comp {N : Type*} [Monoid N] (f : M →* N) (φ : AddChar A M) : AddChar A N :=
```

Comment:
```suggestion
def monoidHomComp {N : Type*} [Monoid N] (f : M →* N) (φ : AddChar A M) : AddChar A N :=
```

Reply:
This doesn't seem to match conventions elsewhere in the library, e.g. [ContinuousAt.comp_continuousWithinAt](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/ContinuousOn.html#ContinuousAt.comp_continuousWithinAt). (There is no object called a `MonoidHomComp`.)

Reply:
(Perhaps this shouldn't be in the `AddChar` namespace at all, but should be called `MonoidHom.comp_addChar`?)